File: Offer/C++/LeetCode121.cc
#include<iostream>
#include<vector>
using namespace std;


class Solution {
public:
    int maxProfit(vector<int>& prices) {
        if( prices.size() == 0){
            return 0;
        }

        int maxPro = 0;
        int minNum = prices[0];

        for(int i = 1; i < prices.size(); i++){
            minNum = min(minNum,prices[i]);
            maxPro = max(prices[i]-minNum,maxPro);
        }
        return maxPro;
    }
};

File: top-interview-questions-easy/DP/LeetCode121.go
package DP

func maxProfit(prices []int) int {
	k := 0
	res := 0
	dp := 0
	for i := 0; i < len(prices); i++ {
		if prices[k] > prices[i] {
			k = i
		}
		res = max(dp, prices[i]-prices[k])
		dp = res
	}
	return res
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
